File: lla2ecef/lla2ecef.cpp
#include <math.h>


struct double3 {
    double x;
    double y;
    double z;
};

/**
 * GPS(WGS84) to ECEF conversion function. Formula refer from: https://www.mathworks.com/help/aeroblks/llatoecefposition.html
 * WGS84 parameter obtained from: https://confluence.qps.nl/qinsy/en/world-geodetic-system-1984-wgs84-29855173.html
 * @param lat: latitude in radians
 * @param lon: longitude in radians
 * @param alt: altitude in meter
 */
void gps2ecef(double lat, double lon, double alt, double3 &ecef){
    const double a = 6378137.0;           // semi-major axis (WGS84) unit: m
    const double f = 1/298.257223563;     // flatening factor
    double latS, rS;                      // Geocentric Latitude, Surface Point Radius
    
    latS = atan(pow((1 - f), 2) * tan(lat));
    rS = sqrt(pow(a, 2) / (1 + ((1 / pow((1 - f), 2) - 1.0) * (pow(sin(latS), 2)))));
        
    ecef.x = rS * cos(latS) * cos(lon) + alt * cos(lat) * cos(lon);          // unit:m
    ecef.y = rS * cos(latS) * sin(lon) + alt * cos(lat) * sin(lon);          // unit:m
    ecef.z = rS * sin(latS) + alt * sin(lat);                                // unit:m
}

File: lla2ecef/lla2ecef.py
import numpy as np

def lla2ecef(lat, lon, alt, model = 'WGS84', f = None, a = None):
    """
	Reference: https://www.mathworks.com/help/aeroblks/llatoecefposition.html
	WGS84 Parameter from: https://confluence.qps.nl/qinsy/en/world-geodetic-system-1984-wgs84-29855173.html
	"""
    if model not in ('WGS84', 'geocentric', 'custom'):
	    raise ValueError('Currently only support WGS84, geocentric, or custom model') 
  
	# default
    if model == 'WGS84':
	    f = np.float64(1/298.257223563)
	    a = np.float64(6378137.0) # m
  
    if model == 'geocentric':
	    f = 0.0
	    if a == None:
		    a = np.float64(6378137.0) # m
      
    if model == 'custom':
	    if f == None:
		    raise ValueError ('Missing kwarg (f: flattening factor) for custom model')
	    elif a == None:
		    raise ValueError ('Missing kwarg (a: semi-major axis) for custom model')
	    elif a == None and f == None:
	    	raise ValueError ('Missing kwarg (f: flattening factor, a: semi-major axis) for custom model')
      
	# math
    lat_r = np.radians(lat) # radians
    lon_r = np.radians(lon) # radians
    lat_s = np.arctan((1-f)**2*np.tan(lat_r)) # radians
    r_s = np.sqrt(a**2/(1+((1/(1-f)**2)-1)*(np.sin(lat_s)**2))) # m
  
    ecef_x = r_s*np.cos(lat_s)*np.cos(lon_r) + alt*np.cos(lat_r)*np.cos(lon_r) # m
    ecef_y = r_s*np.cos(lat_s)*np.sin(lon_r) + alt*np.cos(lat_r)*np.sin(lon_r) # m
    ecef_z = r_s*np.sin(lat_s) + alt*np.sin(lat_r) # m
  
    return ecef_x, ecef_y, ecef_z